source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -57,6 +58,15 @@ SslSocket::SslSocket(Envoy::Ssl::ContextSharedPtr ctx, InitialState state,
     ASSERT(state == InitialState::Server);
     SSL_set_accept_state(ssl_);
   }
+
+  SSL_set_mode(ssl_, SSL_MODE_ASYNC);
+}
+
+SslSocket::~SslSocket() {
+  // If we let the SSL socket be destroyed while there is a pending async SSL operation,
+  // it seems that the callback handler will use already freed memory.
+  while (SSL_waiting_for_async(ssl_))
```

Comment:
Could there be some nicer way to handle this, because we don't really need to wait for any async operations anymore in the socket destructor. Could we ask for any async calls to be cancelled?

Reply:
Hm... I tried to simulate closing a socket with an ongoing async job (by sending SIGINT to Envoy while loaded with requests from k6) and this cycle never ends. Probably because `SSL_waiting_for_async()` consists basically of
```
return SSL->job != NULL;
```

If I remove the cycle then I get segfaults like:
```
#0  ASYNC_WAIT_CTX_get_fd (ctx=<optimized out>, key=0x7ffff58925ae, fd=fd@entry=0x7fffedb1e714, custom_data=custom_data@entry=0x7fffedb1e718)
    at ../crypto/async/async_wait.c:73
#1  0x00007ffff58832d1 in qat_wake_job (job=<optimized out>, jobStatus=<optimized out>) at qat_events.c:289
#2  0x00007ffff57e2f3f in LacPke_MsgCallback () from /usr/lib/libqat_s.so
#3  0x00007ffff57ff642 in adf_user_notify_msgs_poll () from /usr/lib/libqat_s.so
#4  0x00007ffff57fb330 in adf_pollRing () from /usr/lib/libqat_s.so
#5  0x00007ffff57fb64d in icp_adf_pollInstance () from /usr/lib/libqat_s.so
#6  0x00007ffff57f536e in icp_sal_CyPollInstance () from /usr/lib/libqat_s.so
#7  0x00007ffff5883a5a in timer_poll_func (ih=<optimized out>) at qat_polling.c:220
#8  0x00007ffff7ae1fb7 in start_thread (arg=<optimized out>) at pthread_create.c:486
#9  0x00007ffff7a112cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) p curr
$1 = (struct fd_lookup_st *) 0x10064
(gdb) p curr->del
Cannot access memory at address 0x10088
```

The segfault happens even before the callback is able to use already freed memory. 
```c
  66 int ASYNC_WAIT_CTX_get_fd(ASYNC_WAIT_CTX *ctx, const void *key,
  67                           OSSL_ASYNC_FD *fd, void **custom_data)
  68 {
  69     struct fd_lookup_st *curr;
  70 
  71     curr = ctx->fds;
  72     while (curr != NULL) {
  73         if (curr->del) {
  74             /* This one has been marked deleted so do nothing */
  75             curr = curr->next;
  76             continue;
  77         }
  78         if (curr->key == key) {
  79             *fd = curr->fd;
  80             *custom_data = curr->custom_data;
  81             return 1;
  82         }
  83         curr = curr->next;
  84     }
  85     return 0;
  86 }
```
By the time `qat_wake_job()` uses `ctx` it's been already freed when deleting `SslSocket`.

Reply:
Added scheduled deletion of `struct SSL` the way it's done in haproxy.